fedimint-server/src/config/api.rs

```diff
@@ -37,27 +47,157 @@ impl ConfigApi {
     }
 
     // Sets the auth and decryption key derived from the password
-    pub fn set_password(&self, auth: ApiAuth) -> Result<(), ApiError> {
+    pub fn set_password(&self, auth: ApiAuth) -> Result<()> {
         let mut state = self.state.lock().expect("lock poisoned");
 
-        if *state != ConfigApiState::SetPassword {
-            Err(ApiError::bad_request("Password already set".to_string()))
-        } else {
-            *state = ConfigApiState::SetPeerConnectionInfo(auth);
-            Ok(())
+        match &*state {
+            ConfigApiState::SetPassword => *state = ConfigApiState::SetConnections(auth),
+            _ => return Err(ApiError::bad_request("Password already set".to_string())),
+        }
+
+        Ok(())
+    }
+
+    /// Sets our connection info, possibly sending it to a leader
+    pub async fn set_config_gen_connections(
+        &self,
+        request: ConfigGenConnectionsRequest,
+    ) -> Result<()> {
+        let connection = {
+            let mut state = self.state.lock().expect("lock poisoned");
+
+            let connection = match (*state).clone() {
+                ConfigApiState::SetConnections(auth) => ConfigGenConnections::new(request, auth)?,
+                ConfigApiState::SetConfigGenParams(old) => old.with_request(request),
+                _ => return Err(ApiError::bad_request("Config already set".to_string())),
+            };
+            *state = ConfigApiState::SetConfigGenParams(connection.clone());
+
+            connection
+        };
+
+        if let Some(url) = connection.request.leader_api_url.clone() {
+            // Note PeerIds don't really exist at this point, but id doesn't matter because
+            // it's not used in the WS client for anything, perhaps it should be removed
+            let client = WsAuthenticatedClient::new(url, PeerId::from(0), connection.auth.clone());
```

Comment:
This sounds dangerous, why would we send *our* auth data to the leader? They can't verify it and could log into our guardian later? We could have a psk that is required to connect, but that's not what this seems to be and that should likely be a separate PR.

Reply:
Good point, I'll remove the auth from that call.

Reply:
There are a few endpoints @Maan2003 refactored to wait for a certain event to occur, although none of these is subscription-based yet afaik. E.g. waiting for a transaction to be confirmed or rejected:

https://github.com/fedimint/fedimint/blob/cf9917ef806e19890d0073ddb8e9efe7b6c48b65/fedimint-server/src/net/api.rs#L228-L239

You couldn't rely on the DB and would probably need to use [`Notify`](https://dtantsur.github.io/rust-openstack/tokio/sync/struct.Notify.html) directly.